index/faults/faults.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/textileio/powergate/chainstore"
+	"github.com/textileio/powergate/chainsync"
+	"github.com/textileio/powergate/signaler"
+	txndstr "github.com/textileio/powergate/txndstransform"
+	"github.com/textileio/powergate/util"
+	"go.opencensus.io/stats"
+)
+
+const (
+	batchSize = 2000
+)
+
+var (
+	// hOffset is the # of tipsets from the heaviest chain to
+	// consider for index updating; this to reduce sensibility to
+	// chain reorgs
+	hOffset = abi.ChainEpoch(20)
+
+	log = logging.Logger("index-faults")
+)
+
+// Index builds and provides faults history of miners
+type Index struct {
+	api      *apistruct.FullNodeStruct
+	store    *chainstore.Store
+	signaler *signaler.Signaler
+
+	lock  sync.Mutex
+	index IndexSnapshot
+
+	ctx      context.Context
+	cancel   context.CancelFunc
+	finished chan struct{}
+	closed   bool
+}
+
+// New returns a new FaultIndex. It will load previous state from ds, and
+// immediately start getting in sync with new on-chain.
+func New(ds datastore.TxnDatastore, api *apistruct.FullNodeStruct) (*Index, error) {
+	cs := chainsync.New(api)
+	store, err := chainstore.New(txndstr.Wrap(ds, "chainstore"), cs)
+	if err != nil {
+		return nil, err
+	}
+	initMetrics()
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &Index{
+		api:      api,
+		store:    store,
+		signaler: signaler.New(),
+		index: IndexSnapshot{
+			Miners: make(map[string]Faults),
+		},
+		ctx:      ctx,
+		cancel:   cancel,
+		finished: make(chan struct{}),
+	}
+	if err := s.loadFromDS(); err != nil {
+		return nil, err
+	}
+	go s.start()
+	return s, nil
+}
+
+// Get returns a copy of the current index information
+func (s *Index) Get() IndexSnapshot {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	ii := IndexSnapshot{
+		TipSetKey: s.index.TipSetKey,
+		Miners:    make(map[string]Faults, len(s.index.Miners)),
+	}
+	for addr, v := range s.index.Miners {
+		history := make([]int64, len(v.Epochs))
+		copy(history, v.Epochs)
+		ii.Miners[addr] = Faults{
+			Epochs: history,
+		}
+	}
+	return ii
+}
+
+// Listen returns a a signaler channel which signals that index information
+// has been updated.
+func (s *Index) Listen() <-chan struct{} {
+	return s.signaler.Listen()
+}
+
+// Unregister frees a channel from the signaler hub.
+func (s *Index) Unregister(c chan struct{}) {
+	s.signaler.Unregister(c)
+}
+
+// Close closes the FaultIndex.
+func (s *Index) Close() error {
+	log.Info("Closing")
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.closed {
+		return nil
+	}
+	s.cancel()
+	<-s.finished
+	s.closed = true
+	return nil
+}
+
+// start is a long running job that keeps the index up to date with chain updates.
+func (s *Index) start() {
+	defer close(s.finished)
+	if err := s.updateIndex(); err != nil {
+		log.Errorf("initial updating faults index: %s", err)
+	}
+	for {
+		select {
+		case <-s.ctx.Done():
+			log.Info("graceful shutdown of background faults updater")
+			return
+		case <-time.After(util.AvgBlockTime):
```

Comment:
Ah cool, so you are basically polling every `AvgBlockTime`? No listening for chain updates etc